spring-cloud-stream-binder-kstream/src/main/java/org/springframework/cloud/stream/binder/kstream/KStreamBinder.java

```diff
@@ -0,0 +1,184 @@
+															ExtendedProducerProperties<KStreamProducerProperties> properties) {
+		ExtendedProducerProperties<KafkaProducerProperties> extendedProducerProperties = new ExtendedProducerProperties<KafkaProducerProperties>(
+				new KafkaProducerProperties());
+		this.kafkaTopicProvisioner.provisionProducerDestination(name, extendedProducerProperties);
+		if (HeaderMode.embeddedHeaders.equals(properties.getHeaderMode())) {
+			outboundBindTarget = outboundBindTarget.map((k, v) -> {
+				if (v instanceof Message) {
+					try {
+						return new KeyValue<>(k, KStreamBinder.this.serializeAndEmbedHeadersIfApplicable((Message<?>) v));
+					}
+					catch (Exception e) {
+						throw new IllegalArgumentException(e);
+					}
+				}
+				else {
+					throw new IllegalArgumentException("Wrong type of message " + v);
+				}
+			});
+		}
+		else {
+			if (!properties.isUseNativeEncoding()) {
+				outboundBindTarget = outboundBindTarget
+						.map((k, v) -> KeyValue.pair(k, (Object) KStreamBinder.this.serializePayloadIfNecessary((Message<?>) v)));
+			}
+			else {
+				outboundBindTarget = outboundBindTarget
+						.map((k, v) -> KeyValue.pair(k, ((Message<Object>) v).getPayload()));
+			}
+		}
+		if (!properties.isUseNativeEncoding() || StringUtils.hasText(properties.getExtension().getKeySerde()) || StringUtils.hasText(properties.getExtension().getValueSerde())) {
+			try {
+				Serde<?> keySerde;
+				Serde<?> valueSerde;
+
+				if (StringUtils.hasText(properties.getExtension().getKeySerde())) {
+					keySerde = Utils.newInstance(properties.getExtension().getKeySerde(), Serde.class);
+					if (keySerde instanceof Configurable) {
+						((Configurable) keySerde).configure(streamsConfig.originals());
+					}
+				}
+				else {
+					keySerde = this.binderConfigurationProperties.getConfiguration().containsKey("key.serde") ?
+							Utils.newInstance(this.binderConfigurationProperties.getConfiguration().get("key.serde"), Serde.class) : Serdes.ByteArray();
+				}
+
+				if (StringUtils.hasText(properties.getExtension().getValueSerde())) {
+					valueSerde = Utils.newInstance(properties.getExtension().getValueSerde(), Serde.class);
+					if (valueSerde instanceof Configurable) {
+						((Configurable) valueSerde).configure(streamsConfig.originals());
+					}
+				}
+				else {
+					valueSerde = this.binderConfigurationProperties.getConfiguration().containsKey("value.serde") ?
+							Utils.newInstance(this.binderConfigurationProperties.getConfiguration().get("value.serde"), Serde.class) : Serdes.ByteArray();
+				}
+				outboundBindTarget.to((Serde<Object>) keySerde, (Serde<Object>) valueSerde, name);
+			}
+			catch (ClassNotFoundException e) {
+				throw new IllegalStateException("Serde class not found: ", e);
+			}
+
+		}
+		else {
+			outboundBindTarget.to(name);
+		}
+		return new DefaultBinding<>(name, null, outboundBindTarget, null);
+	}
+
+	private Object serializeAndEmbedHeadersIfApplicable(Message<?> message) throws Exception {
```

Comment:
Maybe we should consider to get rid of this as well since we already in Kafka 0.11 mode?